torch/csrc/dynamo/eval_frame.c

```diff
@@ -0,0 +1,578 @@
+  if (extra == SKIP_CODE || (callback == Py_False && extra == NULL)) {
+    DEBUG_TRACE("skip %s", name(frame));
+    return eval_frame_default(tstate, frame, throw_flag);
+  }
+
+  // TODO(jansel): investigate directly using the "fast" representation
+  if (PyFrame_FastToLocalsWithError(frame) < 0) {
+    DEBUG_TRACE("error %s", name(frame));
+    return NULL;
+  }
+
+  // A callback of Py_False indicates "run only" mode, the cache is checked, but
+  // we never compile.
+  if (callback == Py_False) {
+    DEBUG_TRACE("In run only mode %s", name(frame));
+    PyCodeObject* cached_code = lookup(extra, frame->f_locals);
+    if (cached_code != NULL) {
+      // used cached version
+      DEBUG_TRACE("cache hit %s", name(frame));
+      return eval_custom_code(tstate, frame, cached_code, throw_flag);
+    } else {
+      DEBUG_TRACE("cache miss %s", name(frame));
+      return eval_frame_default(tstate, frame, throw_flag);
+    }
+  }
+  DEBUG_CHECK(PyDict_CheckExact(frame->f_locals));
+  DEBUG_CHECK(PyDict_CheckExact(frame->f_globals));
+  DEBUG_CHECK(PyDict_CheckExact(frame->f_builtins));
+
+  // We don't run the current custom_eval_frame behavior for guards.
+  // So we temporarily set the callback to Py_None to drive the correct behavior
+  // in the shim.
+  eval_frame_callback_set(Py_None);
+
+  PyCodeObject* cached_code = lookup(extra, frame->f_locals);
+  if (cached_code != NULL) {
+    // used cached version
+    DEBUG_TRACE("cache hit %s", name(frame));
+    // Re-enable custom behavior
+    eval_frame_callback_set(callback);
+    return eval_custom_code(tstate, frame, cached_code, throw_flag);
+  }
+  // cache miss
+
+  PyObject* result =
+      call_callback(callback, (PyObject*)frame, cache_size(extra));
+  if (result == NULL) {
+    // internal exception, returning here will leak the exception into user code
+    // this is useful for debugging -- but we dont want it to happen outside of
+    // testing
+    return NULL;
+  } else if (result != Py_None) {
+    DEBUG_TRACE("create cache %s", name(frame));
+    extra = create_cache_entry(extra, result);
+    Py_DECREF(result);
+    set_extra(frame->f_code, extra);
+    // Re-enable custom behavior
+    eval_frame_callback_set(callback);
+    return eval_custom_code(tstate, frame, extra->code, throw_flag);
+  } else {
+    DEBUG_TRACE("create skip %s", name(frame));
+    Py_DECREF(result);
+    destroy_cache_entry(extra);
+    set_extra(frame->f_code, SKIP_CODE);
+    // Re-enable custom behavior
+    eval_frame_callback_set(callback);
+    return eval_frame_default(tstate, frame, throw_flag);
+  }
+}
+
+static int active_dynamo_threads = 0;
+
+static PyObject* increment_working_threads(PyThreadState* tstate) {
+  active_dynamo_threads = active_dynamo_threads + 1;
+  if (active_dynamo_threads > 0) {
+    enable_eval_frame_shim(tstate);
+  }
+  Py_RETURN_NONE;
+}
+
+static PyObject* decrement_working_threads(PyThreadState* tstate) {
+  if (active_dynamo_threads > 0) {
```

Comment:
Do you ever decrement without incrementing? Should this be a TORCH_CHECK

Reply:
Agreed.